rest: Use query parameters to control resource loading

stickies-v opened the pull request:
In RESTful APIs, [typically](https://rapidapi.com/blog/api-glossary/parameters/query/) path parameters  (e.g. `/some/unique/resource/`) are used to represent resources, and query parameters (e.g. `?sort=asc`) are used to control how these resources are being loaded through e.g. sorting, pagination, filtering, ...

As first [discussed in #17631](https://github.com/bitcoin/bitcoin/pull/17631#discussion_r733031180), the [current REST api](https://github.com/bitcoin/bitcoin/blob/master/doc/REST-interface.md) contains two endpoints `/headers/` and `/blockfilterheaders/` that rather unexpectedly use path parameters to control how many (filter) headers are returned in the response. While this is no critical issue, it is unintuitive and we are still early enough to easily phase this behaviour out and ensure new endpoints (if any) do not have to stick to non-standard behaviour just for internal consistency.

In this PR, a new `HTTPRequest::GetQueryParameter` method is introduced to easily parse query parameters, as well as two new `/headers/` and `/blockfilterheaders/` endpoints that use a count query parameter are introduced. The old path parameter-based endpoints are kept without too much overhead, but the documentation now points to the new query parameter-based endpoints as the default interface to encourage standardness.

## Behaviour change
### New endpoints and default values
`/headers/` and `/blockfilterheaders/` now have 2 new endpoints that contain query parameters (`?count=<count>`) instead of path parameters (`/<count>/`), as described in REST-interface.md. Since query parameters can easily have default values, I have set this at 5 for both endpoints.

**headers**
`GET /rest/headers/<BLOCK-HASH>.<bin|hex|json>?count=<COUNT=5>`
should now be used instead of
`GET /rest/headers/<COUNT>/<BLOCK-HASH>.<bin|hex|json>`

**blockfilterheaders**
`GET /rest/blockfilterheaders/<FILTERTYPE>/<BLOCK-HASH>.<bin|hex|json>?count=<COUNT=5>`
should now be used instead of
`GET /rest/blockfilterheaders/<FILTERTYPE>/<COUNT>/<BLOCK-HASH>.<bin|hex|json>`

### Some previously invalid API calls are now valid
API calls that contained query strings in the URI could not be parsed prior to this PR. This PR changes behaviour in that previously invalid calls (e.g. `GET /rest/headers/5/somehash.json?someunusedparam=foo`) would now become valid, as the query parameters are properly parsed, and discarded if unused. 
For example, prior to this PR, adding an irrelevant `someparam` parameter would be illegal:
```
GET /rest/headers/5/0000004c6aad0c89c1c060e8e116dcd849e0554935cd78ff9c6a398abeac6eda.json?someparam=true
->
Invalid hash: 0000004c6aad0c89c1c060e8e116dcd849e0554935cd78ff9c6a398abeac6eda.json?someparam=true
```
**This behaviour change affects all rest endpoints, not just the 2 new ones introduced here.**

*(Note: I'd be open to implementing additional logic to refuse requests containing unrecognized query parameters to minimize behaviour change, but for the endpoints that we currently have I don't really see the point for that added complexity. E.g. I don't see any scenarios where misspelling a parameter could lead to harmful outcomes)*

## Using the REST API

To run the API HTTP server, start a bitcoind instance with the `-rest` flag enabled. To use the
`blockfilterheaders` endpoint, you'll also need to set `-blockfilterindex=1`: 
```
./bitcoind -signet -rest -blockfilterindex=1
```

As soon as bitcoind is fully up and running, you should be able to query the API, for example by
using curl on the command line: ```curl "127.0.0.1:38332/rest/chaininfo.json"```. 
To more easily parse the JSON output, you can also use tools like 'jq' or `json_pp`, e.g.:
```
curl -s "localhost:38332/rest/blockfilterheaders/basic/0000004c6aad0c89c1c060e8e116dcd849e0554935cd78ff9c6a398abeac6eda.json?count=2" | json_pp .
```

## To do
- [x] update `doc/release-notes`

## Feedback
This is my first PR (hooray!). Please don't hold back on any feedback/comments/nits/... you may have, big or small, whether they are code, process, language, ... related. I welcome private messages too if there's anything you don't want to clutter the PR with. I'm here to learn and am grateful for everyone's input.

stickies-v committed:
Add GetQueryParameter helper function

Easily get the query parameter from the URI, with optional default value.

stickies-v committed:
Handle query string when parsing data format

URLs may contain a query string (prefixed with '?') and this should be ignored when parsing
the data format.

To facilitate testing this functionality, ParseDataFormat has been made non-static.

stickies-v committed:
Add release notes

stickies-v commented:
Force pushed to rebase after #24054 was merged; no other changes.

brunoerg reviewed (commented):
Concept ACK

Tested ?count=1, ?count=5 and ?count=600 (didn't paste here the result of it, obviously haha) for /rest/headers/
```
http://127.0.0.1:8332/rest/headers/000000001aeae195809d120b5d66a39c83eb48792e068f8ea1fea19d84a4278a.json?count=1

[{"hash":"000000001aeae195809d120b5d66a39c83eb48792e068f8ea1fea19d84a4278a","confirmations":237003,"height":50000,"version":1,"versionHex":"00000001","merkleroot":"27f1d66f8a1ee5280f4e92508dcb647e954d53004905d08a75574daee4988360","time":1270916538,"mediantime":1270915355,"nonce":58489010,"bits":"1c2a1115","difficulty":6.085476906000794,"chainwork":"00000000000000000000000000000000000000000000000000014a426593e15b","nTx":1,"previousblockhash":"000000000845517b31c6820d83f25cff46429bf136a7515fe504116427e60f8e","nextblockhash":"000000001c920d495e1eeef2452b6d1c6c229a919b28196c103ecffebabee141"}]

http://127.0.0.1:8332/rest/headers/000000001aeae195809d120b5d66a39c83eb48792e068f8ea1fea19d84a4278a.json?count=5

[{"hash":"000000001aeae195809d120b5d66a39c83eb48792e068f8ea1fea19d84a4278a","confirmations":240714,"height":50000,"version":1,"versionHex":"00000001","merkleroot":"27f1d66f8a1ee5280f4e92508dcb647e954d53004905d08a75574daee4988360","time":1270916538,"mediantime":1270915355,"nonce":58489010,"bits":"1c2a1115","difficulty":6.085476906000794,"chainwork":"00000000000000000000000000000000000000000000000000014a426593e15b","nTx":1,"previousblockhash":"000000000845517b31c6820d83f25cff46429bf136a7515fe504116427e60f8e","nextblockhash":"000000001c920d495e1eeef2452b6d1c6c229a919b28196c103ecffebabee141"},{"hash":"000000001c920d495e1eeef2452b6d1c6c229a919b28196c103ecffebabee141","confirmations":240713,"height":50001,"version":1,"versionHex":"00000001","merkleroot":"ee3a2d2b895cafacff526d06a55b55e049cf84a9735e4a63f7fd08f96d0f4649","time":1270917100,"mediantime":1270915522,"nonce":56717043,"bits":"1c2a1115","difficulty":6.085476906000794,"chainwork":"00000000000000000000000000000000000000000000000000014a487b7bc7c6","nTx":3,"previousblockhash":"000000001aeae195809d120b5d66a39c83eb48792e068f8ea1fea19d84a4278a","nextblockhash":"000000002066d7f9b134c30b5005b7bf5fbfa52f279883f1fde793dcdc964266"},{"hash":"000000002066d7f9b134c30b5005b7bf5fbfa52f279883f1fde793dcdc964266","confirmations":240712,"height":50002,"version":1,"versionHex":"00000001","merkleroot":"5164dc2785549f9efe14eb1c54522ec1874a02b7eda164fde370c05412f037ad","time":1270917181,"mediantime":1270915874,"nonce":9804838,"bits":"1c2a1115","difficulty":6.085476906000794,"chainwork":"00000000000000000000000000000000000000000000000000014a4e9163ae31","nTx":1,"previousblockhash":"000000001c920d495e1eeef2452b6d1c6c229a919b28196c103ecffebabee141","nextblockhash":"00000000237ce1b692f5ef9f1f9ac64d62d5e6a24dcfec08633fa2d339434e2a"},{"hash":"00000000237ce1b692f5ef9f1f9ac64d62d5e6a24dcfec08633fa2d339434e2a","confirmations":240711,"height":50003,"version":1,"versionHex":"00000001","merkleroot":"792c4a613e51966fc1c852cd8d94fd10a05efa55fffdd6a41281c8d139540b23","time":1270917916,"mediantime":1270916181,"nonce":311792473,"bits":"1c2a1115","difficulty":6.085476906000794,"chainwork":"00000000000000000000000000000000000000000000000000014a54a74b949c","nTx":1,"previousblockhash":"000000002066d7f9b134c30b5005b7bf5fbfa52f279883f1fde793dcdc964266","nextblockhash":"000000000528e8aac90e3096abb7b1f3d52efa2cd8f12754614788cb43eefe11"},{"hash":"000000000528e8aac90e3096abb7b1f3d52efa2cd8f12754614788cb43eefe11","confirmations":240710,"height":50004,"version":1,"versionHex":"00000001","merkleroot":"3cfbd9a91f033e37f05764215626c542c735ed71cd5b53467834cf881c0b9ebd","time":1270918264,"mediantime":1270916354,"nonce":78553962,"bits":"1c2a1115","difficulty":6.085476906000794,"chainwork":"00000000000000000000000000000000000000000000000000014a5abd337b07","nTx":1,"previousblockhash":"00000000237ce1b692f5ef9f1f9ac64d62d5e6a24dcfec08633fa2d339434e2a","nextblockhash":"0000000000e1ad737f11fe4cb3d126408b2cbcc28ab0d83293f408e413733338"}]
```

I tested the same values but using `/rest/headers/<COUNT>/<BLOCK-HASH>.<bin|hex|json>` format and it worked as well. I didn't understand `Deprecated (but not removed) since 23.0:`, will both ways be kept?

MarcoFalke commented:
Is there a large cost to maintaining both ways? If not, it might be better to keep both to not upset users for a weak motivation.

brunoerg commented on src/rest.cpp:
```diff
@@ -195,15 +193,27 @@ static bool rest_headers(const std::any& context,
-    std::string hashStr = path[1];
     uint256 hash;
     if (!ParseHashStr(hashStr, hash))
         return RESTERR(req, HTTP_BAD_REQUEST, "Invalid hash: " + hashStr);
```
if I try to pass a blockhash without the extension (.<bin|hex|json>) in `/headers`, like:
`/rest/headers/000000001aeae195809d120b5d66a39c83eb48792e068f8ea1fea19d84a4278a?count=5`
returns: `Invalid hash: 000000001aeae195809d120b5d66a39c83eb48792e068f8ea1fea19d84a4278a?count=5`

In the deprecated, `/rest/headers/5/000000001aeae195809d120b5d66a39c83eb48792e068f8ea1fea19d84a4278a`
returns: `output format not found (available: .bin, .hex, .json)`

stickies-v replied:
You're right, that should be handled better. I've:
- updated `ParseDataFormat` to ignore the query string even when no data format is found
- updated documentation on `ParseDataFormat` in `rest.h` to document existing and new behaviour
- added a test case to `test_query_string` to capture what you've found

Rebased and pushed to 0415aaafb

brunoerg replied:
Tested! Works!

stickies-v commented:
> Is there a large cost to maintaining both ways? If not, it might be better to keep both to not upset users for a weak motivation.

The PR is already setup to maintain both ways, so anyone using the (to be) deprecated endpoints should not be affected. This is documented in REST-interface.md too, so users are aware it is still supported but that they should probably use the new endpoint for new applications. The maintenance overhead is minimal, basically in [`rest_headers`](https://github.com/bitcoin/bitcoin/pull/24098/files#diff-590507deaf686d6571f73979df5f3c6da6013a13e597f390b8facce39f7c69d1R200-R204) and [`rest_filter_header`](https://github.com/bitcoin/bitcoin/pull/24098/files#diff-590507deaf686d6571f73979df5f3c6da6013a13e597f390b8facce39f7c69d1R372-R376) we have an if/else branch that adds 3 LoC to maintain the old endpoints too. Some additional refactoring could be done to simplify without the two branches, but it's no big deal.

Personally, I would definitely not remove the old endpoints in the next release, and probably not until some bigger API refactoring would make this desirable, in which case we should notify users more explicitly. I did not add any deprecation logging to avoid behaviour change, and because I think we don't have to push users away from the old endpoints yet.

stickies-v commented:
> I didn't understand `Deprecated (but not removed) since 23.0:`, will both ways be kept?

Thank you for the review! Correct, I argue both should be kept at least for now, and that's how it's currently implemented. I'm open to suggestions for better phrasing, but typically deprecated means a feature is no longer developed and may be removed in the future, but is still available. See e.g. [Microsoft's definition](https://docs.microsoft.com/en-us/dynamics365/fin-ops-core/dev-itpro/migration-upgrade/deprecated-features). Does that clear things up?

jsarenik commented:
@stickies-v instead of deprecating the old behavior the new can be just called backwards-compatible. By the way, an example of an externalized backward-compatible REST-service (in a way similar to what you have here) using Caddy webserver can be seen at https://github.com/jsarenik/bitcoin-faucet-shell/blob/master/Caddyfile.txt

jsarenik reviewed (approved):
utACK 0415aaafb2f7206c117f58cc7a4a967cee0fa619

stickies-v commented:
> @stickies-v instead of deprecating the old behavior the new can be just called backwards-compatible.

We can, but I'm not convinced we should? I think deprecating the old endpoints has a clear benefit of keeping the documented interface simpler and better adhering to common REST API practices. Simultaneously, since they're not removed, existing users are not affected, just gently nudged towards the new interface if and when it suits them. 
Thank you for your review!

promag commented on doc/REST-interface.md:
```diff
@@ -47,18 +47,24 @@ The HTTP request and response are both handled entirely in-memory.
 
 #### Blockheaders
-`GET /rest/headers/<COUNT>/<BLOCK-HASH>.<bin|hex|json>`
+`GET /rest/headers/<BLOCK-HASH>.<bin|hex|json>?count=<COUNT=5>`
```
Why only `count` as a query parameter? How about
```
GET /rest/headers?block_hash=&count=&format=

block_hash defaults to tip
count defaults to 5
format defaults to request accept header, JSON if not specified
```

stickies-v replied:
I agree and considered that, but decided to omit this for now to keep this PR's scope limited. Would you be okay with this being incorporated in a follow up PR (which I'm happy to take on right away?)

stickies-v replied:
To conclude, I agree with promag that `/headers/` and `/blockfilterheaders/` are collection endpoints since we're not obtaining a resource identified by the specified `block_hash` but rather use it as a pagination parameter. I didn't properly realise this at first. 

Given that 1) this PR has had quite a bit of review time already which I'd rather not overhaul too much and 2) these and the other updates (moving `format` and `notxdetails` to a query parameter) make sense to implement together, I'll implement this in a follow up PR (I've already briefly started but it will take a bit more time as I'm thinking about ways to minimise the maintenance burden of keeping backwards compatibility).

fjahr replied:
FWIW, I agree with @promag that this would be more properly following REST best practices, which seems to be the main point of this PR.

Do I understand correctly that you want to keep maintaining the old structure, this structure and then also future structures like suggested above @stickies-v ? That does seem more laborious than necessary to me. You could merge these first parts of changes as intermediary steps and then add follow-ups but I would not "officially" support it by adding it to docs and release notes yet. You could make the follow-up changes and only afterwards add documentation and release notes after everything is final, so that there are at least only on two versions to maintain and not one per follow-up PR since without docs and release notes it can be reasonably assumed that nobody is building a client on top of the intermediary version. 

Alternatively if you really want to include docs I would suggest in this intermediate step instead of marking the old structure as deprecated, add the new structure marked as experimental and add a comment that it is still object to change.

stickies-v replied:
I think your approach makes a lot of sense, fjahr. However, since there has already been significant review time spent on this and this PR seems ready for merge, I don't think this is big enough of a problem to ask for another round of review. Even if it's not a huge update, it still would touch quite a few commits (dropping release notes, modifying REST interface docs, updating error return messages and inline docs). I'll definitely consider the approach you suggest for further PRs because I think it is a more careful approach, but for now I'd prefer to leave things as they are and move forward.

promag commented on src/httpserver.cpp:
```diff
@@ -639,6 +640,42 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod() const
+        evhttp_clear_headers(&params_q);
+    }
+
+    if (default_val != "")
```
Maybe use std::optional because an empty string could be the default?

stickies-v replied:
Agreed and done, for both `GetQueryParameterFromUri` and `HTTPRequest:GetQueryParameter`.

promag commented on src/httpserver.cpp:
```diff
@@ -639,6 +640,42 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod() const
+
+std::string GetQueryParameterFromUri(const std::string& uri, const std::string& key, const std::string& default_val)
+{
+    auto query_start = uri.rfind('?');
```
Should use `evhttp_uri_parse`.

promag reviewed (commented):
I don't see a strong reason for this change TBH.

stickies-v commented:
Rebased and:
- fixed wrong indentation of [GetQueryParameter](https://github.com/bitcoin/bitcoin/compare/0415aaafb2f7206c117f58cc7a4a967cee0fa619..e1822e79b3d5230f976360e5500d7ef59dd8783c#diff-aa9c90add56b31a60e0d574fa9025945be2b01306c265d00e3aa0a27c6b4bc3fL132-R146) documentation
- incorporated promag's comments on [std::optional](https://github.com/bitcoin/bitcoin/pull/24098#discussion_r790336836) and [evhttp_uri_parse](https://github.com/bitcoin/bitcoin/pull/24098#discussion_r790338713)

Thank you @promag for your review. To address your comment:
> I don't see a strong reason for this change TBH

I agree this is not a critical issue. However, I think everyone can agree my proposal is the more standard way of organizing API endpoints. If that's true (and so far no one argued the opposite), it's better to refactor this earlier rather than later, because for example in [#17631](https://github.com/bitcoin/bitcoin/pull/17631#discussion_r733031180) we (I think rightfully) preferred internal consistency over adhering to best practices. We now also have the functionality in place for other existing or new endpoints to easily accept query parameters, including your [proposal](https://github.com/bitcoin/bitcoin/pull/24098#discussion_r790336549) for the data format.

_Note: I've just learned I should only rebase in case of merge conflicts. My apologies for the added review difficulty, won't happen going forward._

jnewbery commented on src/httpserver.h:
```diff
@@ -83,6 +84,19 @@ class HTTPRequest
+     *
+     * @param[in] key represents the query parameter of which the value is returned
+     * @param[in] default_val (optional) is returned if key is not in query string
+     * @throws std::runtime_error if key is not in query string and no default_val was provided
```
We don't usually use exceptions to return error conditions in the Bitcoin Core codebase. If you do this, wherever there's a key that doesn't have a default, the calling code will have to wrap this call in a try/except, and then catch the runtime_error and return it as a RESTERR.

Did you consider returning a `std::optional<std::string>`, and then using [std::optional<t>::value_or()](https://en.cppreference.com/w/cpp/utility/optional/value_or) to add the default, and [std::optional<t>::has_value()](https://en.cppreference.com/w/cpp/utility/optional/operator_bool) to catch if a non-default option hasn't been set?

stickies-v replied:
I did not consider that, but that's a great idea. I've taken it a bit further, and removed the `default_val` optional argument from GetQueryParameter, since this functionality can now just as easily be achieved in the headers endpoints by using `std::optional::value_or()` as you suggested, slightly reducing complexity of the code.

jnewbery commented on src/rest.cpp:
```diff
@@ -355,13 +367,30 @@ static bool rest_filter_header(const std::any& context, HTTPRequest* req, const
+        return RESTERR(req, HTTP_BAD_REQUEST, "Invalid URI format. Expected /rest/blockfilterheaders/<filtertype>/<blockhash>.<ext>?count=<count>");
+    }
+
+    auto parsed_count = ToIntegral<size_t>(raw_count);
```
Any reason to lose the `const` here?

stickies-v replied:
Good catch, this is leftover from before I introduced the new `raw_count` variable but indeed should be `const` now. Fixed in both `rest_headers` and `rest_filter_header`.

jnewbery commented on test/functional/interface_rest.py:
```diff
@@ -53,12 +53,13 @@ def skip_test_if_missing_module(self):
-            rest_uri += '.bin'
-        elif req_type == ReqType.HEX:
-            rest_uri += '.hex'
+        query_string = ''
```
Did you consider making `query_string` an optional argument to `test_rest_request()`? It seems a bit contrived to pass the URI without the `req_type` as part of the string, but with the query_string.

stickies-v replied:
I agree, that's a more elegant solution. To facilitate dynamic URI creation, I've added a `query_params` kwargs to `test_rest_request()` which is then encoded into a query string.

jnewbery reviewed (commented):
Concept ACK. A few minor comments inline.

> Why only count as a query parameter? How about
>
> GET /rest/headers?block_hash=&count=&format=
>
> block_hash defaults to tip
> count defaults to 5
> format defaults to request accept header, JSON if not specified

Each `/headers/block` is a unique resource, so I think it makes sense to keep the block hash as part of the URI path. Making `format` a query string key seems ok to me.

I also think that the `/rest/block/notxdetails/` endpoint could be deprecated and merged into `/rest/block/<BLOCKHASH>?txdetails=true|false` or similar in a follow up.

stickies-v commented on src/rest.cpp:
```diff
@@ -14,6 +14,7 @@
 #include <node/context.h>
 #include <primitives/block.h>
 #include <primitives/transaction.h>
+#include <rest.h>
```
Done

stickies-v commented on src/rest.cpp:
```diff
@@ -137,25 +132,30 @@ static ChainstateManager* GetChainman(const std::any& context, HTTPRequest* req)
+    param = strReq.substr(0, std::min(pos_format, pos_querystr));
+
+    // No format string is found
+    if (pos_format == std::string::npos)
```
Done

theStack replied:
This doesn't seem to be resolved?

stickies-v replied:
Sorry, I think I refactored it out again after simplifying `ParseDataFormat` a bit more. Didn't realize our coding style preferred brackets for single-statement multi-line `if` statements. Added brackets again, thanks for catching this.

stickies-v commented on src/rest.cpp:
```diff
@@ -355,13 +367,30 @@ static bool rest_filter_header(const std::any& context, HTTPRequest* req, const
+        // deprecated path: /rest/blockfilterheaders/<filtertype>/<count>/<blockhash>
+        blockhash_index = 2;
+        raw_count = uri_parts[1];
+    }
+    else if (uri_parts.size() == 2) {
```
Done

stickies-v commented on src/rest.cpp:
```diff
@@ -355,13 +367,30 @@ static bool rest_filter_header(const std::any& context, HTTPRequest* req, const
+        // new path with query parameter: /rest/blockfilterheaders/<filtertype>/<blockhash>?count=<count>
+        blockhash_index = 1;
+        raw_count = req->GetQueryParameter("count", "5");
+    }
+    else {
```
Done

stickies-v committed:
Refactoring: move declarations to rest.h

This facilitates unit testing

stickies-v committed:
scripted-diff: rename RetFormat to RESTResponseFormat

As RetFormat is now exposed in a header, it is renamed to the more
understandable RESTResponseFormat
-BEGIN VERIFY SCRIPT-
s() { sed -i 's/RetFormat/RESTResponseFormat/g' $1; }
s src/rest.cpp
s src/rest.h
-END VERIFY SCRIPT-

stickies-v commented on src/rest.h:
```diff
@@ -0,0 +1,28 @@
+
+#include <string>
+
+enum class RetFormat {
```
That's a good idea. I've scripted-diff renamed `RetFormat` to `RESTResponseFormat`, since this enum is currently only used to describe response format.

stickies-v committed:
Update /<count>/ endpoints to use a '?count=' query parameter instead

In most RESTful APIs, path parameters are used to represent resources, and
query parameters are used to control how these resources are being filtered/sorted/...

The old /<count>/ functionality is kept alive to maintain backwards compatibility,
but new paths with query parameters are introduced and documented as the default
interface so future API methods don't break consistency by using query parameters.

stickies-v commented on src/test/rest_tests.cpp:
```diff
@@ -0,0 +1,43 @@
+    BOOST_CHECK_EQUAL(param, "/rest/endpoint/someresource");
+    BOOST_CHECK_EQUAL(rf, RetFormat::JSON);
+
+    // Incorrectly formed query string will not be handled
```
Added an additional test case. More importantly, the handling of this edge case led me to simplify `ParseDataFormat` by refactoring out the `pos_querystr` variable.

stickies-v commented:
Thank you for the extensive review jnewbery, I've incorporated all of your comments into the latest force push 9d9f177. 

> Each `/headers/block` is a unique resource, so I think it makes sense to keep the block hash as part of the URI path. Making `format` a query string key seems ok to me.
>
> I also think that the `/rest/block/notxdetails/` endpoint could be deprecated and merged into `/rest/block/<BLOCKHASH>?txdetails=true|false` or similar in a follow up.

I agree with both of your comments, and am happy to take them on in a follow up PR. Since all of the query parameter logic is already in place now, this should be quite limited in scope too.

promag commented:
> Each `/headers/block` is a unique resource, so I think it makes sense to keep the block hash as part of the URI path.

@jnewbery this endpoint returns a list of `count` headers, not the unique resource.

```
// list of headers
GET /rest/headers?block_hash=&count=&format=

VS

// an header
GET /rest/headers/block_hash?format=
```

Anyway, just wanted to note my opinion, I'm fine with block hash in the path.

stickies-v commented:
Rebased to fix merge conflict from #24223.

LarryRuane commented:
It may be helpful to other reviews to note that you can interact with a local `bitcoind`'s REST interface from the command line using `curl`, for example:
```
$ curl -s localhost:8332/rest/blockhashbyheight/723418.json | jq .
{
  "blockhash": "000000000000000000010ce22c300e67d706860e8e883ed108691a903801b0ee"
}
```
Another common JSON formatting tool is `json_pp` (if you don't have `jq` installed).

gunar commented:
concept ACK

OliverOffing commented:
concept ACK

theStack commented on src/rest.cpp:
```diff
@@ -133,25 +133,28 @@ static ChainstateManager* GetChainman(const std::any& context, HTTPRequest* req)
+        return rf_names[0].rf;
 
+    // Match format string to available formats
+    auto len_suff = param.size() - pos_format - 1;  // - 1 so we don't include the leading .
+    const std::string suff(param, pos_format + 1, len_suff);
```
Calculating the length shouldn't be necessary here, you can simply omit the third parameter of the ctor to get the sub-string from pos_format + 1 to the end:
```suggestion
    const std::string suff(param, pos_format + 1);
```
(I checked that tests still pass after this change)
Also, I'd suggest to rename "suff" to "suffix" to make it more clear (nit though)

stickies-v replied:
Ah yes, you're right, that is more elegant. Updated to your suggestion, and also renamed `suff` to `suffix`.

theStack commented on src/rest.h:
```diff
@@ -14,6 +14,15 @@ enum class RESTResponseFormat {
+ * @param[out]  param       The strReq without the data format string and
+ *                          without the query string (if any).
+ * @param[in]   strReq      The URI to be parsed.
+ * @return      RESTResponseFormat   RESTResponseFormat that was parsed from the URI.
```
nit: no reason to explicitly menton the return type here before the description (we don't do that usually as far as I saw)
```suggestion
 * @return      RESTResponseFormat that was parsed from the URI.
```

theStack commented on src/test/rest_tests.cpp:
```diff
@@ -0,0 +1,48 @@
+    BOOST_CHECK_EQUAL(rf, RESTResponseFormat::JSON);
+
+    // Query string with single parameter
+    rf = ParseDataFormat(param, "/rest/endpoint/someresource.json?p1=v1");
+    BOOST_CHECK_EQUAL(param, "/rest/endpoint/someresource");
+    BOOST_CHECK_EQUAL(rf, RESTResponseFormat::JSON);
```
Could use .bin here (and .hex below, or vice versa) to also test all the different response formats.

stickies-v replied:
Like the idea, incorporated!

stickies-v replied:
Like the idea, incorporated!

theStack commented on doc/release-notes-24098.md:
```diff
@@ -0,0 +1,22 @@
+  For `/blockfilterheaders/`, use
+  `GET /rest/blockfilterheaders/<FILTERTYPE>/<BLOCK-HASH>.<bin|hex|json>?count=<COUNT=5>`
+  instead of
+  `GET /rest/blockfilterheaders/<FILTERTYPE>/<COUNT>/<BLOCK-HASH>.<bin|hex|json>`
```
nit: to be consistent with the headers description above:
```suggestion
  `GET /rest/blockfilterheaders/<FILTERTYPE>/<COUNT>/<BLOCK-HASH>.<bin|hex|json> (deprecated)`
```

stickies-v replied:
Thanks, fixed

glozow reviewed (commented):
Concept ACK, I don't know much about REST but am convinced that query parameter instead of path makes more sense for specifying count.

theStack commented:
@stickies-v: Seems like you didn't force-push your latest changes yet?

w0xlt reviewed (commented):
Approach ACK. 
I agree that a query parameter is more adequate for the `count` parameter.
Also agree on the refactoring and the test added.

I think the commit 833803e can be split in two: one for the change mentioned in the title of  commit and other for the test added.

stickies-v commented:
Force pushed changes to incorporate suggestions from theStack and an re-unresolved comment from jnewbery. As per my [comment](https://github.com/bitcoin/bitcoin/pull/24098#discussion_r815467602) in the thread, I think the best approach is to move the `block_hash` into the follow up PR, so I think for now all comments on the content of this PR are resolved.

> It may be helpful to other reviews to note that you can interact with a local `bitcoind`'s REST interface from the command line using `curl`, for example:

Thanks for the suggestion, I updated the initial post with some instructions.

> @stickies-v: Seems like you didn't force-push your latest changes yet?

My bad, I tended to use the "Mark as resolved" as a personal to do list which I now realize is confusing, going forward I'll try to force push first and comment/mark afterwards.

> I think the commit 833803e can be split in two: one for the change mentioned in the title of commit and other for the test added.

Could you please explain how that would improve the PR in your view? I think since the test is quite small, so closely linked with the code changes, and wouldn't work without the prior PR, it's quite intuitive to keep it in one? I had a look at other PRs and I couldn't really find consensus between doing tests in the same or separate commits, so I just went for my intuition.

theStack reviewed (commented):
Left some more review comments, mostly code-style nit-picking though, feel free to ignore.
Changes generally LGTM and I think the PR is pretty close to a mergeable state.

stickies-v commented on src/httpserver.cpp:
```diff
@@ -639,6 +641,39 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod() const
+
+std::optional<std::string> GetQueryParameterFromUri(const char* uri, const std::string& key)
+{
+    auto uri_parsed = evhttp_uri_parse(uri);
```
I hadn't considered this but I agree - updated this and in `HTTPRequest::GetQueryParameter`.

stickies-v commented on src/httpserver.cpp:
```diff
@@ -639,6 +641,39 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod() const
+{
+    auto uri_parsed = evhttp_uri_parse(uri);
+    auto query = evhttp_uri_get_query(uri_parsed);
+    std::optional<std::string> result {};
```
Thanks, I wasn't aware of this convention. I've updated this and 3 other zero-inits of `std::string`. I'm still zero-init'ing `size_t blockhash_index {};` as to my understanding `long` doesn't have a default ctor and this could be unsafe otherwise.

stickies-v commented:
> Left some more review comments, mostly code-style nit-picking though, feel free to ignore. Changes generally LGTM and I think the PR is pretty close to a mergeable state.

Thank you for the thorough review, I've incorporated all of your suggestions. Hoping to get some final ACKs to get this over the line, I agree that it feels like being close to merged 👍

jnewbery commented on src/httpserver.cpp:
```diff
@@ -639,6 +641,38 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod() const
+                break;
+            }
+        }
+        evhttp_clear_headers(&params_q);
```
Are you sure this `evhttp_clear_headers()` call is necessary? The `evkeyvalq` struct is destructed immediately after this.

jnewbery replied:
Yes, that looks correct to me. Thanks for looking into this!

jnewbery commented on src/httpserver.cpp:
```diff
@@ -639,6 +641,38 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod() const
+
+    if (query) {
+        // Parse the query string into a key-value queue and iterate over it
+        struct evkeyvalq params_q;
```
Is the `struct` keyword needed here?

```suggestion
       evkeyvalq params_q;
```

stickies-v replied:
This confused me too. I don't see how it's necessary to keep it, but that's how libevent [documents](https://libevent.org/doc/http_8h_source.html) it and also how it's already used everywhere ([1](https://github.com/bitcoin/bitcoin/blob/767ee2e3a1082468b4e2248bac3ef8bd54bb2ddb/src/httpserver.cpp#L525), [2](https://github.com/bitcoin/bitcoin/blob/767ee2e3a1082468b4e2248bac3ef8bd54bb2ddb/src/httpserver.cpp#L556), [3](https://github.com/bitcoin/bitcoin/blob/3c565302aaa91ae46aa09b8aeb95206711d9d8a6/src/bitcoin-cli.cpp#L766)) in Bitcoin Core, as well as in most (all?) unofficial examples/documentation you can find online. I thought it might be a good idea to keep it to stay in line with libevent conventions. Knowing that, would you still prefer to have it removed? I don't have a strong view either way, but think I prefer sticking with the convention.

jnewbery replied:
ok, I've dug into this a bit more, and it's an [elaborated type specifier](https://en.cppreference.com/w/cpp/language/elaborated_type_specifier). It means that even if we have a local variable named `evkeyvalq`, then this would still compile, since the `struct` class-key would mean the local variable is ignored.

I think it's fine to leave it.

jnewbery commented on src/httpserver.cpp:
```diff
@@ -639,6 +641,38 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod() const
+        struct evkeyvalq params_q;
+        evhttp_parse_query_str(query, &params_q);
+
+        struct evkeyval* param;
+        for (param = params_q.tqh_first; param; param = param->next.tqe_next) {
```
I'd suggest a couple of changes:

- put the `param` declaration in the for loop init-statement
- explicitly compare `param` with `nullptr`:

```suggestion
        for (evkeyval* param{params_q.tqh_first}; param != nullptr; param = param->next.tqe_next) {
```

stickies-v replied:
Done - although as per my other [comment](https://github.com/bitcoin/bitcoin/pull/24098#discussion_r816865897) I've kept the `struct` prefix for now to stick with convention.

jnewbery commented on src/rest.cpp:
```diff
@@ -351,17 +360,32 @@ static bool rest_filter_header(const std::any& context, HTTPRequest* req, const
-    if (uri_parts.size() != 3) {
-        return RESTERR(req, HTTP_BAD_REQUEST, "Invalid URI format. Expected /rest/blockfilterheaders/<filtertype>/<count>/<blockhash>");
+    std::string raw_count;
+    size_t blockhash_index {};
```
What do you think about handling the blockhash in the same way as the count:

```diff
--- a/src/rest.cpp
+++ b/src/rest.cpp
@@ -365,14 +365,14 @@ static bool rest_filter_header(const std::any& context, HTTPRequest* req, const
     std::vector<std::string> uri_parts;
     boost::split(uri_parts, param, boost::is_any_of("/"));
     std::string raw_count;
-    size_t blockhash_index {};
+    std::string raw_blockhash;
     if (uri_parts.size() == 3) {
         // deprecated path: /rest/blockfilterheaders/<filtertype>/<count>/<blockhash>
-        blockhash_index = 2;
+        raw_blockhash = uri_parts[2];
         raw_count = uri_parts[1];
     } else if (uri_parts.size() == 2) {
         // new path with query parameter: /rest/blockfilterheaders/<filtertype>/<blockhash>?count=<count>
-        blockhash_index = 1;
+        raw_blockhash = uri_parts[1];
         raw_count = req->GetQueryParameter("count").value_or("5");
     } else {
         return RESTERR(req, HTTP_BAD_REQUEST, "Invalid URI format. Expected /rest/blockfilterheaders/<filtertype>/<blockhash>.<ext>?count=<count>");
@@ -384,8 +384,8 @@ static bool rest_filter_header(const std::any& context, HTTPRequest* req, const
     }
 
     uint256 block_hash;
-    if (!ParseHashStr(uri_parts[blockhash_index], block_hash)) {
-        return RESTERR(req, HTTP_BAD_REQUEST, "Invalid hash: " + uri_parts[blockhash_index]);
+    if (!ParseHashStr(raw_blockhash, block_hash)) {
+        return RESTERR(req, HTTP_BAD_REQUEST, "Invalid hash: " + raw_blockhash);
     }
 
     BlockFilterType filtertype;
```

theStack reviewed (approved):
ACK fe95d30ae4ffd1946823ec3e2d6a59c063d62893 🌱

jnewbery commented on src/rest.cpp:
```diff
@@ -191,19 +190,29 @@ static bool rest_headers(const std::any& context,
+        return RESTERR(req, HTTP_BAD_REQUEST, "Invalid URI format. Expected /rest/headers/<hash>.<ext>?count=<count>");
+    }
+
+    const auto parsed_count = ToIntegral<size_t>(raw_count);
```
Is there a reason you've changed `parsed_count` from using braced initialization to using `=` initialization?

stickies-v replied:
I don't think this was an intentional change. I've changed both `parsed_count` definitions to use brace initialization.  I've also updated all other new/modified `=` initializations I could find to use brace initialization. Specifically, this is in `ParseDataFormat`, `HTTPRequest::GetQueryParameter` and `GetQueryParameterFromUri`

jnewbery commented on src/rest.cpp:
```diff
@@ -351,17 +360,32 @@ static bool rest_filter_header(const std::any& context, HTTPRequest* req, const
+        return RESTERR(req, HTTP_BAD_REQUEST, "Invalid URI format. Expected /rest/blockfilterheaders/<filtertype>/<blockhash>.<ext>?count=<count>");
+    }
+
+    const auto parsed_count = ToIntegral<size_t>(raw_count);
```
Same question about changing from braced initialization to `=` initialization.

stickies-v replied:
Fixed

jnewbery reviewed (commented):
nearly ACK fe95d30ae4ffd1946823ec3e2d6a59c063d62893

I've left a few minor comments. The only one that I'd definitely like to see resolved before merge is the build failure on the first two commits.

stickies-v commented on src/httpserver.cpp:
```diff
@@ -639,6 +641,37 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod() const
+        struct evkeyvalq params_q;
+        evhttp_parse_query_str(query, &params_q);
+
+        for (struct evkeyval* param { params_q.tqh_first }; param != nullptr; param = param->next.tqe_next) {
```
Fixed, wasn't aware of this convention, thanks. I really do appreciate the nits, consistent style is important.

stickies-v commented on src/rest.cpp:
```diff
@@ -137,25 +133,28 @@ static ChainstateManager* GetChainman(const std::any& context, HTTPRequest* req)
 }
 
-static RetFormat ParseDataFormat(std::string& param, const std::string& strReq)
+RESTResponseFormat ParseDataFormat(std::string& param, const std::string& strReq)
```
Yikes, this shouldn't have happened, thanks for catching that. I've adapted [this](https://blog.ploeh.dk/2013/10/07/verifying-every-single-commit-in-a-git-branch/) script to ensure all commits in the branch build:

```sh
BRANCH=$(git rev-parse --abbrev-ref HEAD)
COMMITS=$(git log master..$BRANCH --oneline  --reverse | cut -d " " -f 1)

for COMMIT in $COMMITS
do
  git -c advice.detachedHead=false checkout $COMMIT > /dev/null
  make -j "$(($(sysctl -n hw.physicalcpu)+1))" > /dev/null  # on linux, use: make -j "$(($(nproc)+1))" > /dev/null
  if [ $? -eq 0 ]
  then
        echo $COMMIT - passed
    else
        echo $COMMIT - failed
        git checkout $BRANCH
        exit
    fi
done
echo all commits on $BRANCH built successfully

git checkout $BRANCH
```

MarcoFalke replied:
You don't need a script for that. It is built-in to git already, see https://github.com/bitcoin/bitcoin/blob/master/doc/productivity.md#interactive-dummy-rebases-for-fixups-and-execs-with-git-merge-base

stickies-v commented:
Force pushed from fe95d30 to 9aac438 to address jnewbery's comments, namely [not each commit being able to compile](https://github.com/bitcoin/bitcoin/pull/24098#discussion_r822575408) and several style updates regarding spaces and brace initialization. There are no functional/behaviour changes.

The range-diff (`git range-diff 219d728 fe95d30 9aac438`) is a bit noisy unfortunately, you can also compare the commits individually:
```
git diff 1cc759af212c91807735e6c9bc2c998468c7252a..9f1c54787c81177dd56a31c881a9ad2834a122dc
git diff 40830ae7ea72c6ac6082bbfd30d26d1f749c46df..c1aad1b3b95b7c6bdf05e0c2095aba2f2db8310b
git diff 2c84d3916b710fbf6029858acff941549e875cb7..fff771ee864975cee8c831651239bac95503c37a
git diff ac3c007a393d3493ad9023c71d4ed3c38a8baabd..a09497614e9bb603fff36286d9611a25b23eeb02
git diff f610294c449aefc4bdae9175292fedbe5dd1d5f5..46af138c216d5fa7ac512a42c8219706e97d88fa
git diff fe95d30ae4ffd1946823ec3e2d6a59c063d62893..9aac4381897130ce630c4c3f5fca6690cd148330
```

jnewbery commented:
Code review ACK 9aac4381897130ce630c4c3f5fca6690cd148330

Tested range-diff and that all commits build.

theStack reviewed (approved):
re-ACK 9aac4381897130ce630c4c3f5fca6690cd148330

stickies-v commented:
I've had to push a tiny doc update to `REST-interface.md` (`git range-diff 219d728 9aac438 54b39cf`) since this was not merged for v23, but since there are no significant changes beyond theStack and jnewbery's ACKs I think this PR is now ready to be considered for merging? @MarcoFalke 

<details>
    <summary>git range-diff 219d728 9aac438 54b39cf</summary>

```
1:  9f1c54787 = 1:  9f1c54787 Refactoring: move declarations to rest.h
2:  c1aad1b3b = 2:  c1aad1b3b scripted-diff: rename RetFormat to RESTResponseFormat
3:  fff771ee8 = 3:  fff771ee8 Handle query string when parsing data format
4:  a09497614 = 4:  a09497614 Add GetQueryParameter helper function
5:  46af138c2 ! 5:  f959fc039 Update /<count>/ endpoints to use a '?count=' query parameter instead
    @@ doc/REST-interface.md: The HTTP request and response are both handled entirely i
      Given a block hash: returns <COUNT> amount of blockheaders in upward direction.
      Returns empty if the block doesn't exist or it isn't in the active chain.
      
    -+*Deprecated (but not removed) since 23.0:*
    ++*Deprecated (but not removed) since 24.0:*
     +`GET /rest/headers/<COUNT>/<BLOCK-HASH>.<bin|hex|json>`
     +
      #### Blockfilter Headers
    @@ doc/REST-interface.md: The HTTP request and response are both handled entirely i
      direction for the filter type <FILTERTYPE>.
      Returns empty if the block doesn't exist or it isn't in the active chain.
      
    -+*Deprecated (but not removed) since 23.0:*
    ++*Deprecated (but not removed) since 24.0:*
     +`GET /rest/blockfilterheaders/<FILTERTYPE>/<COUNT>/<BLOCK-HASH>.<bin|hex|json>`
     +
      #### Blockfilters
6:  9aac43818 = 6:  54b39cfb3 Add release notes
```
</details>

theStack reviewed (approved):
re-ACK 54b39cfb342d10a448d49299c715e3a25c2aca4a

Verified that the only change since my last ACK was bumping the version number in the REST documentation from v23 to v24 (since this PR didn't make it into v23, v24 will be the first release where the old endpoints are deprecated).

jnewbery commented:
ACK 54b39cfb342d10a448d49299c715e3a25c2aca4a
